docs/hugo/content/design/secrets.md

```diff
@@ -0,0 +1,669 @@
+
+Conclusion: TBD
+
+**What's the difference between `x-ms-secret` and `format: password`?**
+
+Unclear currently. I have a question out to the Swagger team. Current plan is to just treat them the same.
+
+Conclusion: TBD
+
+### Azure generated secrets
+
+Azure generated secrets will be _optionally_ downloaded to a Kubernetes or KeyVault secret. Users instruct the operator to download
+the secrets associated with a resource by supplying a `SecretDestination` in the `Spec`.
+
+The optionality of this step is key for [preferring managed identity](#prefer-managed-identity), as it allows those in AAD/Managed identity cases to avoid
+secrets they don't want/won't use being put into their namespace.
+
+```go
+type SecretDestination struct {
+    // Secret includes details for where to store this secret in Kubernetes.
+    Kubernetes *KubernetesSecretDestination  // TODO: We could call this "Secret" instead
+
+    // KeyVault includes details for where to store this secret in KeyVault.
+    KeyVault *KeyVaultSecretDestination
+}
+
+type KubernetesSecretDestination struct {
+    // Name is the name of the secret.
+    Name string
+
+    // Annotations will be added to the created secrets annotations.
+    Annotations map[string]string
+}
+
+type KeyVaultSecretDestination struct {
+    // Reference refers to the KeyVault to write the secret to.
+    Reference ResourceReference
+
+    // Name is the name of the secret to create in KeyVault.
+    Name string
+}
+```
+
+Using this in the YAML would look like this, for an example CosmosDB `DatabaseAccount`:
+
+```yaml
+spec:
+  # Other spec fields elided...
+  forOperator:
+    secrets:
+      # Save the read-write keys (and endpoints) into a kubernetes secret called "my-secret" and a KeyVault secret called "my-secret".
+      keyDestination:
+        kubernetes:
+          name: my-secret
+        keyVault:
+          reference:
+            armId: /subscriptions/.../resourceGroups/.../providers/Microsoft.KeyVault/vaults/asokeyvault
+          name: my-secret
+      # Save the read only keys just to a Kubernetes secret called "my-readonly-secret"
+      readOnlyKeyDestination:
+        kubernetes:
+          name: my-readonly-secret
+```
+
+Note that some resources (like CosmosDB `DatabaseAccount`) have multiple kinds of secrets. There might be a `PrimaryKey`, `SecondaryKey`, `ReadOnlyPrimaryKey`, 
+and `ReadOnlySecondaryKey`. At the very least we need to support putting the main keys and the readonly keys into two different secrets. To accomplish this,
+we can define multiple logical secret groupings in the ASO config and a corresponding `destination` property will be created for each of them. You can see this
+done above in the sample.
+
+Some fields such as `Endpoint` (or `AccountName` for other databases) are not _really_ secret, but should be included in these logical secret groups anyway so
+that they're easier for users to inject into pods.
```

Comment:
In a lot of ways, yes, it conflicts.
I've added this as an open question so we can discuss further.

> I'm also cautious about classifying things as secretes merely to allow them to be injected into pods.

The issue is that values like `endpoint` are generated by Azure, and there isn't really a way to get those values injected into pods _except_ putting them into a `configMap` or `secret`. We _could_ make a `configMap` with the non-secret values and a `secret` with the secret values but that seems more burdensome/confusing from an operations perspective.

We shouldn't just make the users copy the returned value, as there's no good way to do that in a gitops style workflow.

This is in contrast to _user specified_ values of this sort, which they can just duplicate into 2 places in their `spec`'s - one in the `DB` `AccountName` field (for example) and 1 in a secret of their choice (for use in injecting into their pods if they desire). Since they control this value they can use client tooling/gitops to manage it. It's a bit icky still since the same value is in two places but it's at least doable.

Reply:
Fair enough.